src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/rc_pending_resumption_handler.cc

```diff
@@ -0,0 +1,292 @@
+    LOG4CXX_DEBUG(logger_,
+                  "Freezed request with function id: "
+                      << fid << " and correlation_id: " << cid);
+  }
+
+  for (auto subscription : need_to_subscribe) {
+    // request_ref = CreateResumptionRequest(subscription)
+    // const uint32_t corr_id =
+    //      request_ref[application_manager::strings::params]
+    //                 [application_manager::strings::correlation_id]
+    //                    .asUInt();
+    // auto resumption_request = MakeResumptionRequest(corr_id, function_id,
+    // *request); save (resumption_request) subscribe_on_event(function_id,
+    // corr_id); subscriber(app.app_id(), resumption_request);
+    // rpc_service_.ManageHMICommand(request);
+    pending_subscriptions_.push_back(subscription);
+    const auto cid = application_manager_.GetNextHMICorrelationID();
+    auto subscription_request = CreateSubscriptionRequest(subscription, cid);
+    auto fid = GetFunctionId(subscription_request);
+    auto resumption_request =
+        MakeResumptionRequest(cid, fid, subscription_request);
+    pending_requests_.emplace(cid, *subscription_request);
+    LOG4CXX_DEBUG(logger_,
+                  "Subscribing for event with function id: "
+                      << fid << " correlation id: " << cid);
+    subscribe_on_event(fid, cid);
+    subscriber(app_id, resumption_request);
+    LOG4CXX_DEBUG(logger_,
+                  "Sending request with function id: "
+                      << fid << " and correlation_id: " << cid);
+    application_manager_.GetRPCService().ManageHMICommand(subscription_request);
+  }
+}
+
+void RCPendingResumptionHandler::ClearPendingResumptionRequests() {
+  // remove all freazed subscriptions
+  // remove saved resumption requeasts
+  LOG4CXX_AUTO_TRACE(logger_);
+  sync_primitives::AutoLock lock(pending_resumption_lock_);
+
+  for (auto const& it : pending_requests_) {
+    const auto timed_out_pending_request_fid =
+        static_cast<hmi_apis::FunctionID::eType>(
+            it.second[app_mngr::strings::params][app_mngr::strings::function_id]
+                .asInt());
+    unsubscribe_from_event(timed_out_pending_request_fid);
+  }
+
+  pending_requests_.clear();
+  freezed_resumptions_.clear();
+  pending_subscriptions_.clear();
+}
+
+void RCPendingResumptionHandler::ProcessSuccessfulResponse(
+    const application_manager::event_engine::Event& event,
+    const ModuleUid& module_uid) {
+  LOG4CXX_AUTO_TRACE(logger_);
+
+  auto& response = event.smart_object();
+  auto cid = event.smart_object_correlation_id();
+
+  RaiseEvent4Response(response, cid);
+
+  unsubscribe_from_event(event.id());
+
+  const auto& it_queue_freezed = freezed_resumptions_.find(module_uid);
+  if (it_queue_freezed != freezed_resumptions_.end()) {
+    LOG4CXX_DEBUG(logger_, "Freezed resumptions found");
+    auto& queue_freezed = it_queue_freezed->second;
+    while (!queue_freezed.empty()) {
+      const auto& resumption_request = queue_freezed.front();
```

Comment:
Why do we rize only for the first in queue? 
 suppose that we should raize all freezed applications that response from HMI is available